metadata-ingestion/src/datahub/ingestion/source/file.py

```diff
@@ -178,34 +180,19 @@ def __init__(self, ctx: PipelineContext, config: FileSourceConfig):
         self.ctx = ctx
         self.config = config
         self.report = FileSourceReport()
-        self.fp: Optional[BufferedReader] = None
+        self.fp = None
 
     @classmethod
     def create(cls, config_dict, ctx):
         config = FileSourceConfig.parse_obj(config_dict)
         return cls(ctx, config)
 
-    def get_filenames(self) -> Iterable[str]:
-        path_parsed = parse.urlparse(str(self.config.path))
-        if path_parsed.scheme in ("file", ""):
-            path = pathlib.Path(self.config.path)
-            if path.is_file():
-                self.report.total_num_files = 1
-                return [str(path)]
-            elif path.is_dir():
-                files_and_stats = [
-                    (str(x), os.path.getsize(x))
-                    for x in path.glob(f"*{self.config.file_extension}")
-                    if x.is_file()
-                ]
-                self.report.total_num_files = len(files_and_stats)
-                self.report.total_bytes_on_disk = sum([y for (x, y) in files_and_stats])
-                return [x for (x, y) in files_and_stats]
-            else:
-                raise Exception(f"Failed to process {path}")
-        else:
-            self.report.total_num_files = 1
-            return [str(self.config.path)]
+    def get_filenames(self) -> Iterable[FileStatus]:
+        path_str = str(self.config.path)
+        fs = FileSystem.get(path_str)
```

Comment:
this overall feels a bit complex

is there a reason we can't use smart_open directly without building wrappers for each thing (e.g. http, local, s3)?

Reply:
If I am not mistaken **smart_open** does not have methods to list path and get file info. It can open stream only. Currently there are 3 file systems: **local**, **s3** and **http**. In the future, there can be more. And if someone need to add azure support, for instance, the only one thing that should be done is to implement FileSystem for Azure. Thats it.